src/librustc_resolve/lib.rs

```diff
@@ -172,13 +172,16 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver,
                                    resolution_error: ResolutionError<'a>)
                                    -> DiagnosticBuilder<'sess> {
     match resolution_error {
-        ResolutionError::TypeParametersFromOuterFunction => {
+        ResolutionError::TypeParametersFromOuterFunction(outer_span) => {
```

Comment:
if we have the full `Def` here, we can match on it and see whether it is `Def::SelfType` or some other kind of def. If it is a self-type, we can customize the error message to highlight the trait/impl. The tricky bit is still getting the *span* from a `Def`.